src/server.c

```diff
@@ -5318,48 +5319,44 @@ int main(int argc, char **argv) {
                 exit(1);
             }
         }
-
-        /* First argument is the config file name? */
-        if (argv[j][0] != '-' || argv[j][1] != '-') {
-            configfile = argv[j];
-            server.configfile = getAbsolutePath(configfile);
-            /* Replace the config file in server.exec_argv with
-             * its absolute path. */
-            zfree(server.exec_argv[j]);
-            server.exec_argv[j] = zstrdup(server.configfile);
-            j++;
+        /* Parse command line options
+         * Precedence wise, File, stdin, explicit options -- last config is the one that matters.
+         *
+         * First argument is the config file name? */
+        if (argv[1][0] != '-') {
+            /* Replace the config file in server.exec_argv with its absolute path. */
+            server.configfile = getAbsolutePath(argv[1]);
+            zfree(server.exec_argv[1]);
+            server.exec_argv[1] = zstrdup(server.configfile);
+            j = 2; // Skip this arg when parsing options
+        }
+        /* Last argument - Should we read config from stdin? */
+        if (argv[argc-1][0] == '-' && argv[argc-1][1] == '\0') {
+            config_from_stdin = 1;
         }
-
         /* All the other options are parsed and conceptually appended to the
          * configuration file. For instance --port 6380 will generate the
-         * string "port 6380\n" to be parsed after the actual file name
-         * is parsed, if any. */
-        while(j != argc) {
+         * string "port 6380\n" to be parsed after the actual config file
+         * and stdin input are parsed (if they exist). */
+        while(j < (argc - (config_from_stdin ? 1 : 0))) {
             if (argv[j][0] == '-' && argv[j][1] == '-') {
                 /* Option name */
-                if (!strcmp(argv[j], "--check-rdb")) {
-                    /* Argument has no options, need to skip for parsing. */
-                    j++;
-                    continue;
-                }
                 if (sdslen(options)) options = sdscat(options,"\n");
                 options = sdscat(options,argv[j]+2);
-                options = sdscat(options," ");
             } else {
                 /* Option argument */
                 options = sdscatrepr(options,argv[j],strlen(argv[j]));
-                options = sdscat(options," ");
             }
+            options = sdscat(options," ");
             j++;
         }
-        if (server.sentinel_mode && configfile && *configfile == '-') {
-            serverLog(LL_WARNING,
-                "Sentinel config from STDIN not allowed.");
```

Comment:
i suppose there's a reason for that.
probably that sentinel must be able to rewrite it's config and restart.

Reply:
Yes, It still won't start without a configfile -- I just simplified the check a bit since I now have an explicit flag for stdin and don't have to use the hack of saving '-' as the config file name.

Reply:
i think it's not just that it shouldn't start without a config file.
i think the intention was that it should explicitly fail when `config_from_stdin` is true.
if the main reason for providing config via stdin is that we don't store passwords on disk, sentinel isn't designed to do that.

Reply:
Don't think so -- It's perfectly fine to have the config come from stdin initially AS LONG as you also have a file to write to later on.
(it's less useful obviously, but not something that will fail later on).

Reply:
ok. sounds reasonable to me, but i'd like @yossigo to confirm that.